wicket/src/update.rs

```diff
@@ -0,0 +1,241 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Types related to rack updates
+
+use semver::Version;
+use serde::{Deserialize, Serialize};
+use sha3::Digest;
+use sha3::Sha3_256;
+use std::fs::File;
+use std::io::prelude::*;
+use std::path::{Path, PathBuf};
+
+#[derive(Debug, Clone)]
+pub struct Sha3_256Digest([u8; 32]);
+
+impl Serialize for Sha3_256Digest {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        hex::serde::serialize(self.0, serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for Sha3_256Digest {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        Ok(Sha3_256Digest(hex::serde::deserialize(deserializer)?))
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum ArtifactType {
+    // Sled Artifacts
+    SledSp,
+    SledRoT,
+    HostPhase1,
+    HostPhase2,
+
+    // PSC Artifacts
+    PscSp,
+    PscRot,
+
+    // Switch Artifacts
+    SwitchSp,
+    SwitchRot,
+}
+
+/// A description of a software artifact that can be installed on the rack
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Artifact {
+    filename: PathBuf,
+    artifact_type: ArtifactType,
+    version: Version,
+    digest: Sha3_256Digest,
+    length: u64,
+}
+
+/// Attempt to convert an ArtifactSpec to an Artifact
+/// by reading from the filesysetm and hashing.
+impl TryFrom<ArtifactSpec> for Artifact {
+    type Error = std::io::Error;
+    fn try_from(spec: ArtifactSpec) -> Result<Self, Self::Error> {
+        let mut hasher = Sha3_256::new();
+        let mut file = File::open(&spec.filename)?;
```

Comment:
It's probably worth moving beyond `std::io::Error` - this will give the frustrating "file not found" (without naming the file that wasn't found) error message. It looks like `std::io::Error` is used a lot in this file; I'd be fine with merging this as-is with a followup issue to come back and add error types with more context.